File: 897. Increasing Order Search Tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
  TreeNode* dfs(TreeNode* current, TreeNode* last){
    if (current == nullptr){
      return last;
    }
    TreeNode* result = dfs(current->left, current);
    current->left = nullptr;
    current->right = dfs(current->right, last);
    return result;
  }
public:
  TreeNode* increasingBST(TreeNode* root) {
    return dfs(root, nullptr);
  }
};

File: 897. Increasing Order Search Tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var increasingBST = function(root) {
  function dfs(current, last){
    if (current == null){
      return last;
    }
    let result = dfs(current.left, current);
    current.left = null;
    current.right = dfs(current.right, last);
    return result;
  }
  
  return dfs(root, null);
};